pyanaconda/core/string.py

```diff
@@ -0,0 +1,176 @@
+#
+# string.py - generic string utility functions
+#
+# Copyright (C) 2021  Red Hat, Inc.  All rights reserved.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+import string  # pylint: disable=deprecated-module
+import sys
+import unicodedata
+
+
+def strip_accents(s):
+    """This function takes arbitrary unicode string
+    and returns it with all the diacritics removed.
+
+    :param s: arbitrary string
+    :type s: str
+
+    :return: s with diacritics removed
+    :rtype: str
+
+    """
+    return ''.join(c for c in unicodedata.normalize('NFD', s)
+                   if unicodedata.category(c) != 'Mn')
+
+
+def ensure_str(str_or_bytes, keep_none=True):
```

Comment:
I think we can get rid of `ensure_str` and always use `decode_bytes`. The function is always called with `keep_none=True`, so we just need to handle `None`. The `ProxyString` class can have its own private method, `have_word_match` already handles `None` values and `_toASCII ` seems to handle it as well.

Please, don't forget to check all related projects.

Reply:
It's a mess. It appears `decode_bytes` was added in parallel to `ensure_str,`, which makes some sense. The difference is handling of `None` and crucially, encoding used on bytes. So reconciling these two is a bit of apples and oranges in my opinion.

The fact that `ensure_str` is hidden in a number of calls means there are many places that can stealthily take `bytes` and make assumptions about encoding. I'll try to make sense of these first.

Reply:
I think I eliminated `have_word_match`. That leaves two others, `upper_ascii` and `lower_ascii`. A thorough review of their usage (full info below) leaves only `SimpleConfigFile`, which uses this call for dict keys, so I think it can have a private method same as `ProxyString`. Then we could probably eliminate `ensure_str` outright. That would also mean I can drop the commit adding None support to `decode_bytes`, there would be nowhere left to require that.

However, I am not sure if that is all helpful, actually. This all looks to me as an exercise in reshuffling things, while none of the actual code benefits in a tangible way. The benefit to humans is one less function to read, in a file of 175 lines. That's laughable, IMO.

---

So, it's like this for `upper_ascii`:
- `QueueFactory` in `queuefactory.py` is ok, ultimately just str.
- `SimpleConfigFile` in `simpleconfig.py` reads from file and needs conversion from default.
That's somewhat fine.

For `lower_ascii` it's more fun:
- `_normalize_checksum` in `installation.py` - str from pykickstart
- `_complete_device_factory_request` in `add_device.py` - str via D-Bus
- `QueueFactory` in `queuefactory.py` - str, same as for `upper_ascii` above
- `_set_mount_point` in `custom_storage_helpers.py` - str from Gtk

Reply:
See below, it is all deleted.